module/zfs/mmp.c

```diff
@@ -0,0 +1,548 @@
+
+	vd = spa->spa_root_vdev;
+	while (!vd->vdev_ops->vdev_op_leaf) {
+		child = vd->vdev_child[spa_get_random(vd->vdev_children)];
+
+		if (!vdev_writeable(child))
+			continue;
+
+		if (child->vdev_ops->vdev_op_leaf && child->vdev_mmp_pending) {
+			if (pending_writes-- > 0)
+				continue;
+			else
+				return (NULL);
+		}
+
+		vd = child;
+	}
+	return (vd);
+}
+
+static void
+mmp_write_done(zio_t *zio)
+{
+	vdev_t *vd = zio->io_vd;
+	mmp_thread_state_t *mts = zio->io_private;
+
+	mutex_enter(&mts->mmp_io_lock);
+	vd->vdev_mmp_pending = 0;
+
+	if (zio->io_error)
+		goto unlock;
+
+	/*
+	 * Mmp writes are queued on a fixed schedule, but under many
+	 * circumstances, such as a busy device or faulty hardware,
+	 * the writes will complete at variable, much longer,
+	 * intervals.  In these cases, another node checking for
+	 * activity must wait longer to account for these delays.
+	 * 
+	 * Mmp_delay captures the interval between completion of the
+	 * prior mmp write, and the one that just finished.  This
+	 * tells the importer how soon it can expect to see activity
+	 * if the pool is still imported.
+	 *
+	 * Do not set mmp_delay if zfs_mmp_interval == 0, so we do
+	 * not trigger an activity check on import.
+	 */
+	if (zfs_mmp_interval) {
+		hrtime_t delay = gethrtime() - mts->mmp_last_write;
+
+		if (delay > mts->mmp_delay)
+			mts->mmp_delay = delay;
+		else
+			mts->mmp_delay = (delay + mts->mmp_delay * 127) /
+			    128;
+	} else {
+		mts->mmp_delay = 0;
+	}
+	mts->mmp_last_write = gethrtime();
+
+unlock:
+	mutex_exit(&mts->mmp_io_lock);
+}
+
+/*
+ * When the uberblock on-disk is updated by a spa_sync,
+ * creating a new "best" uberblock, update the one stored
+ * in the mmp thread state, used for mmp writes.
+ */
+void
+mmp_update_uberblock(dsl_pool_t *dp, uberblock_t *ub)
+{
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+
+	mutex_enter(&mmp->mmp_io_lock);
+	mmp->mmp_ub = *ub;
+	mmp->mmp_ub.ub_timestamp = gethrestime_sec();
+	mutex_exit(&mmp->mmp_io_lock);
+}
+
+/*
+ * Choose a random vdev, label, and MMP block, and write over it
+ * with a copy of the last-synced uberblock, whose timestamp
+ * has been updated to reflect that the pool is in use.
+ */
+static void
+mmp_write_uberblock(dsl_pool_t *dp, uberblock_t *ub)
+{
+	int flags = ZIO_FLAG_CONFIG_WRITER | ZIO_FLAG_CANFAIL;
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+	spa_t *spa = dp->dp_spa;
+	uberblock_t mmp_ub;
+	vdev_t *vd;
+	uint64_t offset;
+
+	vd = vdev_random_leaf(spa);
+	if (vd == NULL || !vdev_writeable(vd))
+		return;
+
+	mutex_enter(&mmp->mmp_io_lock);
+	mmp_ub = *ub;
+	mmp_ub.ub_timestamp = gethrestime_sec();
+	mmp_ub.ub_mmp_magic = MMP_MAGIC;
+	mmp_ub.ub_mmp_delay = mmp->mmp_delay;
+	vd->vdev_mmp_pending = gethrtime();
+	mutex_exit(&mmp->mmp_io_lock);
+
+	zio_t *zio  = zio_root(spa, NULL, NULL, flags);
+	abd_t *ub_abd = abd_alloc_for_io(VDEV_UBERBLOCK_SIZE(vd), B_TRUE);
+	abd_zero(ub_abd, VDEV_UBERBLOCK_SIZE(vd));
+	abd_copy_from_buf(ub_abd, &mmp_ub, sizeof (uberblock_t));
```

Comment:
Pulling this whole block under the mutex would let us avoid making another local uberblock_t mmp_ub copy.